lib/core/pool/pool.js

```diff
@@ -0,0 +1,289 @@
+const ConnectionCheckedOutEvent = events.ConnectionCheckedOutEvent;
+const ConnectionCheckedInEvent = events.ConnectionCheckedInEvent;
+const PoolClearedEvent = events.PoolClearedEvent;
+
+const VALID_OPTIONS = [
+  'maxPoolSize',
+  'minPoolSize',
+  'maxIdleTimeMS',
+  'waitQueueTimeoutMS',
+  'enableConnectionMonitoring'
+];
+
+function getSpecOptions(options) {
+  const newOptions = VALID_OPTIONS.reduce((obj, key) => {
+    if (options.hasOwnProperty(key)) {
+      obj[key] = options[key];
+    }
+    return obj;
+  }, {});
+
+  return Object.freeze(newOptions);
+}
+
+class Pool extends EventEmitter {
+  constructor(options) {
+    super();
+    options = options || {};
+
+    this.options = getSpecOptions(options);
+
+    const counter = makeCounter(1);
+    const connections = new ConnectionManager();
+    const waitQueue = new WaitQueue({
+      pool: this,
+      waitQueueTimeoutMS:
+        typeof options.waitQueueTimeoutMS === 'number' ? options.waitQueueTimeoutMS : 0
+    });
+
+    this.s = {
+      // Wait queue that handles queueing for connections
+      waitQueue,
+
+      // Connection Manager that handles state of various connections
+      connections,
+
+      // Counter that increments for each new connection.
+      counter,
+
+      // Spec mandated fields
+      maxPoolSize: typeof options.maxPoolSize === 'number' ? options.maxPoolSize : 100,
+      minPoolSize: typeof options.minPoolSize === 'number' ? options.minPoolSize : 0,
+      maxIdleTimeMS: typeof options.maxIdleTimeMS === 'number' ? options.maxIdleTimeMS : 0,
+
+      // Allows us to override the Connection constructor for testing purposes
+      Connection: options.Connection || Connection,
+
+      // State variables that do not fall into any other category
+      pid: process.pid,
+      generation: 0,
+      isClosed: false,
+      address: options.address
+    };
+
+    process.nextTick(() => {
+      this._emitMonitoringEvent(PoolCreatedEvent);
+      this._satisfyMinPoolSize();
+    });
+  }
+
+  // Public API
+  checkOut(callback) {
+    this._emitMonitoringEvent(ConnectionCheckOutStarted);
+
+    if (this.s.isClosed) {
+      this._emitMonitoringEvent(ConnectionCheckOutFailed, 'poolClosed');
+      return callback(new PoolClosedError(this));
+    }
+
+    const self = this;
+
+    this.s.waitQueue.enter(function() {
+      const args = [callback].concat(Array.from(arguments));
+      self._acquisitionHandler.apply(self, args);
+    });
+  }
+
+  checkIn(connection, force, callback) {
+    if (typeof force === 'function' && typeof callback !== 'function') {
+      callback = force;
+      force = false;
+    }
+
+    if (!this.s.connections.has(connection)) {
+      return callback(new PoolReleaseForeignConnectionError(this, connection));
+    }
+
+    const closed = this.s.isClosed;
+    const stale = this._connectionIsStale(connection);
+    const willDestroy = !!(force || closed || stale);
```

Comment:
nit: a little hard to read. Maybe `Boolean(force || closed || stale)` instead?

Reply:
In general, it is not considered good practice to use the "primitive wrapper classes" in JavaScript, as they are not the same thing as the primitives themselves and lead to funky behavior:

```
> x = new Boolean(true)
[Boolean: true]
> typeof x
'object'
> x == true
true
> x === true
false
```

`!!` is considered the conventional way of ensuring that a value is a boolean.